File: java/DoublyLinkedList.java
class Node {
  int data;
  Node next, prev;

  public Node(int data) {
    this.data = data;
    this.next = null;
    this.prev = null;
  }
}

public class DoublyLinkedList {
  Node head;

  public DoublyLinkedList() {
    head = null;
  }

  // add new node at beginning
  public void addFirst(int data) {
    Node newNode = new Node(data);
    if(head == null) {
      head = newNode;
    } else {
      newNode.next = head;
      head.prev = newNode;
      head = newNode;
    }
  }

  // add new node at end
  public void addLast(int data) {
    Node newNode = new Node(data);
    if(head == null) {
      head = newNode;
    } else {
      Node last = head;
      while(last.next != null) {
        last = last.next;
      }
      last.next = newNode;
      newNode.prev = last;
    }
  }

// delete node by key
public void deleteByKey(int key) {
  Node current = head;
  if(current == null) {
    return;
  }
  if(current.data == key) {
    head = current.next;
    if(head != null) {
      head.prev = null;
    }
    return;
  }
  while(current!= null && current.data != key) {
    current = current.next;
  }
  if(current == null) {
    return;
  }
  current.prev.next = current.next;
  if(current.next != null) {
    current.next.prev = current.prev;
  }
}
// print list 
public void printList() {
  Node current = head;
  while(current != null) {
    System.out.print(current.data + " ");
    current = current.next;
  }
  System.out.println();
}
}

File: javascript/doublyLinkedList.js
class Node {
  constructor(data) {
    this.data = data;
    this.prev = null;
    this.next = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }
  isEmpty() {
    return this.head === null;
  }
  addToHead(data) {
    const newNode = new Node(data);
    if(this.isEmpty()) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.next = this.head;
      this.head.prev = newNode;
      this.head = newNode;
    }
  }

  addToTail(data) {
    const newNode = new Node(data);
    if(this.isEmpty()) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      newNode.prev = this.tail;
      this.tail = newNode;
    }
  }

  deleteHead() {
    if(this.isEmpty()) {
      return;
    }
    if(this.head === this.tail) {
      this.head = null;
      this.tail = null;
    } else {
      this.head = this.head.next;
      this.head.prev = null;
    }
  }

  deleteTail() {
    if(this.isEmpty()) {
      return;
    }
    if(this.head === this.tail) {
      this.head = null;
      this.tail = null;
    } else {
      this.tail = this.tail.prev;
      this.tail.next = null;
    }
  }

  printList() {
    if(this.isEmpty()) {
      console.log("list is empty");
      return;
    }
    let current = this.head;
    while(current) {
      console.log(current.data);
      current = current.next;
    }
  }
}
